### PR TITLE
[WIP] Add SSL/TLS Config for Postgres

### PR DESCRIPTION
**Pull request by Chanakya-TS**
# Add SSL/TLS Support for PostgreSQL in embedmd Package

## Description

This PR implements SSL/TLS support for PostgreSQL database connections in the embedmd package, bringing feature parity with the existing MySQL SSL/TLS implementation while accounting for PostgreSQL-specific capabilities and limitations.

### Key Changes Made:

**1. Enhanced PostgreSQL Connector (`internal/datastore/embedmd/postgres/connect.go`)**
- Added `TLSConfig` field to `PostgresDBConnector` struct
- Updated `NewPostgresDBConnector` constructor to accept TLS configuration parameter
- Implemented retry logic with exponential backoff (25 attempts over ~5 minutes)
- Added comprehensive SSL/TLS DSN building methods:
  - `BuildDSNWithTLS()` - Main method for building SSL-enabled DSNs
  - `buildURLDSNWithTLS()` - Handles URL format DSNs (`postgres://...`)
  - `buildKeyValueDSNWithTLS()` - Handles key=value format DSNs
  - `addOrUpdateDSNParam()` - Helper for DSN parameter manipulation

**2. SSL/TLS Configuration Support**
- **SSL Modes**: 
  - `verify-full` when `VerifyServerCert` is true
  - `require` when certificates are provided
- **Certificate Support**:
  - Client certificates via `sslcert` and `sslkey` parameters
  - Root/CA certificates via `sslrootcert` parameter
  - Supports both `RootCertPath` and `CAPath` from TLS config
- **DSN Format Support**:
  - URL format: `postgres://user:pass@host:port/db?params`
  - Key-value format: `host=localhost port=5432 user=postgres dbname=test`

**3. PostgreSQL-Specific Behavior**
- Added warning logging for cipher suite configuration (not supported by PostgreSQL's libpq)
- Proper URL encoding handling for certificate file paths in URL format DSNs
- Graceful handling of unsupported TLS features

**4. Updated Integration Points**
- `internal/db/connector.go`: Updated to pass TLS configuration to PostgreSQL connector
- `internal/datastore/embedmd/postgres/migrate_test.go`: Updated constructor calls

**5. Comprehensive Test Suite (`internal/datastore/embedmd/postgres/connect_test.go`)**
- **Connection Tests**: Basic, SSL, and URL format connections
- **TLS Validation**: Certificate validation, cipher suite warnings, error handling
- **DSN Building**: Both URL and key-value formats with various SSL configurations
- **Secure Connections**: Multiple SSL modes and certificate configurations
- **Error Cases**: Invalid URLs, non-existent certificates, malformed certificates

### **Automated Testing**
- **Unit Tests**: Comprehensive test suite with 20+ test cases covering:
  - SSL/TLS configuration validation
  - DSN building for both URL and key-value formats
  - Certificate path handling and validation
  - Error scenarios and edge cases
  - PostgreSQL-specific behavior (cipher warnings)

- **Integration Tests**: Using testcontainers with real PostgreSQL instances:
  - Basic insecure connections
  - SSL-enabled connections with various modes
  - Certificate-based authentication scenarios

### **Manual Testing**
- Verified DSN building produces correct SSL parameters
- Confirmed certificate file paths are properly URL-encoded in URL format
- Tested warning logging for unsupported cipher configuration
- Validated retry mechanism behavior
- Ensured backward compatibility with existing non-SSL configurations

### **Test Results**
```bash
=== RUN   TestPostgresDBConnector_Connect_Insecure
--- PASS: TestPostgresDBConnector_Connect_Insecure (2.04s)
=== RUN   TestPostgresDBConnector_TLSConfigValidation  
--- PASS: TestPostgresDBConnector_TLSConfigValidation (0.25s)
=== RUN   TestPostgresDBConnector_DSNBuilding
--- PASS: TestPostgresDBConnector_DSNBuilding (0.41s)
=== RUN   TestPostgresDBConnector_Connect_Secure
--- PASS: TestPostgresDBConnector_Connect_Secure (1.83s)
=== RUN   TestPostgresDBConnector_Connect_ErrorCases
--- PASS: TestPostgresDBConnector_Connect_ErrorCases (0.00s)
=== RUN   TestMigrations
--- PASS: TestMigrations (1.61s)
=== RUN   TestDownMigrations  
--- PASS: TestDownMigrations (1.58s)
PASS
ok      github.com/kubeflow/model-registry/internal/datastore/embedmd/postgres  7.757s
```

### **Key Differences from MySQL Implementation**
- **No TLS Registration**: PostgreSQL uses DSN parameters instead of separate TLS config registration
- **Limited Cipher Support**: PostgreSQL doesn't support client-side cipher configuration through connection parameters
- **SSL Mode Parameter**: Uses PostgreSQL's native `sslmode` parameter for SSL behavior control

## Merge criteria:
- [x] All the commits have been [_signed-off_](https://github.com/kubeflow/community/tree/master/dco-signoff-hook#signing-off-commits)  (To pass the `DCO` check)
- [x] The commits have meaningful messages
- [x] Automated tests are provided as part of the PR for major new functionalities; testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work.
- [x] Code changes follow the [kubeflow contribution guidelines](https://www.kubeflow.org/docs/about/contributing/).

### **Supported SSL/TLS Flags**
✅ `--embedmd-database-ssl-verify-server-cert` - Implemented via `sslmode=verify-full`  
⚠️ `--embedmd-database-ssl-cipher` - Not supported by PostgreSQL (warning logged)

This implementation brings PostgreSQL SSL/TLS support to feature parity with MySQL while respecting PostgreSQL's specific capabilities and providing comprehensive test coverage.

### CODE REVIEW

File: internal/db/connector.go [hunk 1]
```diff
@@ -32,8 +32,9 @@ func Init(dbType string, dsn string, tlsConfig *tls.TLSConfig) error {
 	switch dbType {
 	case "mysql":
 		_connectorInstance = mysql.NewMySQLDBConnector(dsn, tlsConfig)
+		_connectorInstance = mysql.NewMySQLDBConnector(dsn, tlsConfig)
```
Comment (Al-Pragliola) [inline internal/db/connector.go]: I think this line here is coming from the rebase

**Review by Al-Pragliola — CHANGES_REQUESTED** [review]

File: internal/db/connector.go (hunk 1 above)
Comment (Chanakya-TS) [reply internal/db/connector.go]: Thank you for catching that! I just  fixed it.

**Review by Al-Pragliola — APPROVED** [review]
/lgtm

**Comment by Al-Pragliola** [comment]
- https://github.com/kubeflow/model-registry/pull/1250#pullrequestreview-3002438495
- https://github.com/kubeflow/model-registry/pull/1250#pullrequestreview-3005688065

/approve
